packages/@ourworldindata/grapher/src/captionedChart/StaticChartRasterizer.tsx

```diff
@@ -0,0 +1,173 @@
+                    ),
+                    testUrl = await urlFromBlob(testBlob),
+                    testImage = await imageFromUrl(testUrl)
+
+                canvas.width = 10
+                canvas.height = 10
+                ctx.drawImage(testImage, 0, 0)
+
+                // ensure that all the pixels have been painted opaque white; if the font hasn't loaded yet,
+                // nothing will be drawn and the pixels will all be transparent black
+                const pixels = ctx.getImageData(0, 0, 10, 10)?.data ?? [],
+                    didDraw = pixels.length && pixels.every((ch) => ch === 255)
+                outcomes.push(!!didDraw)
+            }
+
+            // check that all the fonts rendered correctly and exit if so
+            if (outcomes.every((result) => result)) return
+
+            // otherwise, keep retrying until we reach the time limit
+            console.warn(
+                "preloading fonts...",
+                faces.filter((f, i) => !outcomes[i]).map((f) => f.font)
+            )
+            await new Promise((res) =>
+                setTimeout(res, RETRY_DELAY_MS + (retry * RETRY_DELAY_MS) / 2)
+            )
+        }
+
+        // if the function didn't return from within the loop, the embedded fonts aren't usable
+        // so don't try to insert them when rendering the chart
+        this.embeddedFonts = ""
+    }
+
+    async render(): Promise<{
+        url: string
+        blob: Blob
+        svgUrl: string
+        svgBlob: Blob
+    }> {
+        // await the canvas first before anything else to make sure .preloadFonts() has completed
+        const canvas = await this.canvas
+
+        // create an Image object using the chart's svg, but with embedded fonts swapped in for
+        // its external `fonts.css` stylesheet
+        const svgImage = await imageFromUrl(
+            await urlFromBlob(
+                blobFromSVG(
+                    this.svg.replace(IMPORT_FONTS_REGEX, this.embeddedFonts)
+                )
+            )
+        )
+
+        // render the bitmap version
+        const { width, height, density, format } = this
+        canvas.width = width * density
+        canvas.height = height * density
+        const ctx = canvas.getContext("2d", {
+            alpha: false,
+        }) as CanvasRenderingContext2D
+        ctx.imageSmoothingEnabled = false
+        ctx.setTransform(density, 0, 0, density, 0, 0)
+        ctx.drawImage(svgImage, 0, 0)
+
+        // package the svg version
+        const svgBlob = blobFromSVG(this.svg),
+            svgUrl = await urlFromBlob(svgBlob)
+
+        return new Promise((res, rej) => {
+            try {
+                const url = canvas.toDataURL(format)
+                canvas.toBlob((blob) => {
+                    if (blob) res({ url, blob, svgUrl, svgBlob })
+                    else rej(new Error("Failed to generate bitmap blob"))
+                })
+            } catch (err) {
+                rej(err)
```

Comment:
It would be nice if in the case where the png generation fails for whatever reason, we could still display the svg version.

But only do this if it's easy to do, no need to go down another massive rabbit hole.